### DISCUSSION TITLE
Responses types

### DISCUSSION THREAD

**Discussion by Nihisil** [fastapi/fastapi#7010] [Questions] [answered]
According to documentation this is correct code:
```python
default_responses = {
    status.HTTP_401_UNAUTHORIZED: {
        "description": "Incorrect auth credentials"
    }
}

app = FastAPI(responses=default_responses)
```

But if you lint it with mypy, result will be: 
> Argument "responses" to "FastAPI" has incompatible type "Dict[int, Dict[str, str]]"; expected "Optional[Dict[Union[int, str], Dict[str, Any]]]"

I'm not sure where the issue and how to better fix it:
1. mypy error?
2. not correct types
3. not correct input

**Comment by ycd** [comment, marked as answer]
Not correct type, `status.HTTP_401_UNAUTHORIZED` is `Literal[401]`. 

But in `responses` field we use this signature `responses: Optional[Dict[Union[int, str], Dict[str, Any]]] = None`

The change looks `Literal[int]` simple but it is not possible per [PEP 586](https://www.python.org/dev/peps/pep-0586/). 

If you want to make MyPy happy, as shown in the docs, you need to use it like this.

```python
default_responses = {
    401: {
        "description": "Incorrect auth credentials"
    }
}
```

**Comment by Nihisil** [comment]
Got it. Thank you for the response.

**Comment by ycd** [comment]
You are welcome! If your problem is resolved feel free to close the issue.

**Comment by tiangolo** [comment]
Thanks for the help here @ycd ! 👏  🙇 

Thanks for reporting back and closing the issue @Nihisil  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by major-mayer** [comment]
For some reason, this doesn't work for me:

```python
default_responses = {
    401: {
        "description": "Incorrect auth credentials"
    }
}


@router.put("/v1/[...]/result", response_class=Response, status_code=status.HTTP_204_NO_CONTENT,
            callbacks=get_fastapi_callback_routes(None), responses=default_responses)
```

Mypy (0.991) reports the following
```
src/[...]/compute_feature.py:39:68: error: Argument "responses" to "put" of "APIRouter" has incompatible type "Dict[int, Dict[str, str]]"; expected "Optional[Dict[Union[int, str], Dict[str, Any]]]"  [arg-type]
```

**Reply by dshadow** [reply]
I got the same issue with Mypy 1.2.0 and FastAPI 0.95.1

**Comment by juliancoffee** [comment]
https://github.com/fastapi/fastapi/discussions/16259

I opened a new one, which is relevant, but about a different thing.
Original issue is about covariance over key of respone dict (simply put, making `mypy` accept an integer-like key where it wants `int | str` as a dict key).
And I'm not aware of any type in the stdlib that would work here, and it would probably require a separate type.
Mine is about covariance over value of dict (simply put, making `mypy` accept a more precise response type).

This type would also not be a purely typing change, since it would probably ban a few operations, such as getting an element by key, since it would be unsound.
